game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/aa/AaFireAndCasualtyStep.java

```diff
@@ -26,25 +33,75 @@
 
   @Override
   public List<String> getNames() {
-    final List<String> steps = new ArrayList<>();
-    if (!valid()) {
-      return steps;
-    }
-    for (final String typeAa : UnitAttachment.getAllOfTypeAas(aaGuns())) {
-      steps.add(firingPlayer().getName() + " " + typeAa + AA_GUNS_FIRE_SUFFIX);
-      steps.add(firedAtPlayer().getName() + SELECT_PREFIX + typeAa + CASUALTIES_SUFFIX);
-      steps.add(firedAtPlayer().getName() + REMOVE_PREFIX + typeAa + CASUALTIES_SUFFIX);
-    }
-    return steps;
+    return getSteps().stream()
+        .flatMap(step -> step.getNames().stream())
+        .collect(Collectors.toList());
+  }
+
+  @Override
+  public void execute(final ExecutionStack stack, final IDelegateBridge bridge) {
+    final List<BattleStep> steps = getSteps();
+
+    // steps go in reverse order on the stack
+    Collections.reverse(steps);
+    steps.forEach(stack::push);
   }
 
-  protected boolean valid() {
-    return !aaGuns().isEmpty();
+  private List<BattleStep> getSteps() {
+    return FireStepsBuilder.buildSteps(
+        FireStepsBuilder.Parameters.builder()
+            .battleState(battleState)
+            .battleActions(battleActions)
+            .firingGroupFilter(FiringGroupFilterAa.of(getSide()))
+            .side(getSide())
+            .returnFire(MustFightBattle.ReturnFire.ALL)
+            .roll(new Roll())
+            .selectCasualties(new Casualties())
```

Comment:
Is select casualties an in/out parameter? You probably can use `@builder.default` on 'select casualties' to give it a default value, and avoid passing a new instance.

Similar question about the `new roll()`